### ISSUE TITLE
APIV2: NGSI types Ontologies

### ISSUE THREAD

**Issue by tobiasjacobs** [telefonicaid/fiware-orion#1516]
NEC proposes to include in the NGSI 2 specs syntax and semantics for NGSI type ontologies. 

An NGSI type ontology expresses
-   Which entity types exist in an NGSI system
-   Which value types are used by an NGSI system
-   Which metadata names are used in an NGSI system

Relations:
o   Which entity types have which attributes, and what are the value types of these attributes
o   Which entity types are related by a subtype-relationship. Entity Subtypes inherit all attributes from their supertypes
o   Which metadata names have which metadata types
o   (maybe) which value types can have which metadata

This kind of information can be expressed by an ontology. What we propose is to
-   Define an ontology structure, i.e. specify how the above information is to be expressed
-   Define a lightweight NGSI base ontology containing
  o   Only the entity type “entity” which is the supertype of all other entities
  o   Basic value types for string, numbers, dates/time, maybe geo-location
  o   Basic metadata like timestamp, accuracy, provider
  o   Basic attributes of entity type “entity” for describing e.g. the entity location

A concrete NGSI system deployment can then extend the base ontology with all the entities/attributes/metadata used within the system. By publishing that ontology, users then know which entity types they can query. 

See also the presentation slides by my colleague Martin Bauer and the corresponding example ontology.

[NGSI Context Interfaces and Semantics.pptx](https://github.com/telefonicaid/fiware-orion/files/39167/NGSI.Context.Interfaces.and.Semantics.pptx)

**Comment by tobiasjacobs** [comment]
And here is the example ontology file:

[SmartSantanderNGSI-RDF.owl.txt](https://github.com/telefonicaid/fiware-orion/files/39171/SmartSantanderNGSI-RDF.owl.txt)

**Comment by GillesPrivat** [comment]
basically agreed, but this should be discussed together with the use of JSON-LD, where all references to ontologies would be regrouped under a JSON-LD" @context", which would require minimal change to the current format andi interfaces already published (as existing identifiers can be matched to ontolgy-referencing URIs in the @ context)

**Comment by tobiasjacobs** [comment]
@GillesPrivat, I have read some of the JSON-LD documentation, and it makes a lot of sense to me to use those concepts.

**Comment by jmcanterafonseca** [comment]
I think the proposed Ontology activity has many synergies with an activity we have just started around the creation of standard NGSI data models for different domains, particularly smart cities, taking advantage of the traction of the OASC initiative. A preliminary work can be found here for Smart Parking domain:

https://docs.google.com/document/d/17leIlKCE5EdOtrAurbIsvbjRnE6UMEXQcNVswvS0J_A/edit?usp=sharing

Our aspiration is to publish smart city, smartagri, etc. data models as https://schema.org did for the Web. A catalogue of common metadata (accuracy, timestamp, resolution, etc.) will make sense as well. 

If we want to express data models as ontologies, as well, as @GillesPrivat, said JSON-LD has many advantages as someone can discover what are the semantics of the data by inspecting the context and following their nose to an ontology published somewhere else.

**Comment by GillesPrivat** [comment]
The definition of domain ontologies such as for smart cities is important, but can be agreed on in the next stage. What we need first is to agree on a transversal ontology redefining formally what is and entity  vs a device, and all the identifiers used in the NGSI data model .  Even before this we need to agree on whether to generalize the use of JSON-LD, because JSON-LD provides a very powerful and elegant solution to solve all our problems : if we have redefined all the keywords of the existing NGSI data model in a proper and complete JSON-LD @context, we can keep these existing keywords as aliases ( even keep id instead @id, type instead of@type). This means we could still use existing JSON descriptions for backward compatibility, provided we reference this @context in an HTTP link header with http://www.w3.org/ns/json-ld#context  (in this case you need not even change the media type from JSON to JSON-LD).

**Comment by jmcanterafonseca** [comment]
an ontology of the NGSIv2 data model can be a nice to have and a
contribution being made by the community. But the work regarding data
models for domains can proceed in parallel. All in all we can create data
models not expressed as an ontology per se although they can be later
 expressed using an ontology as well.

2015-11-25 10:55 GMT+01:00 Gilles Privat <notifications@github.com>:

**Comment by FanoRamparany** [comment]
It seems to me that there are 2 separate issues here:
- the need for ontologies for modelling the concepts of the domains of application
- the use of JSON-LD as a format for echanging data about any domain of application
As Jose Manuel says, they are complementary but independant and could be investigated in parallel.

As for the ontologies issue, I don't really understand the need for defining an ontology for the  NGSI data model. NGSI is not a domain of application and NGSI should be able to contain data from any application domain. A more standard approach for me would be to define ways to refer to existing ontologies when formating NGSI messages. For example in the NGSI entity-property-value format, when expressing that "John leavesIn Paris":
- "John", "Paris" should have a URI, and be instances of classes defined in one the ontologies.
- these ontologies should be declared somewhere in the NGSI message  (for example in the meta-data part) or as a preamble message.
- "leavesIn" should have an URI and be a property defined in one of the ontologies.

What worries me in your proposal when you say:
"An NGSI type ontology expresses which entity types exist in an NGSI system"
is that it seems to limit the applicability of the system. What if I'd like to tell something about an entity which type is not in the NGSI type ontology? But maybe I didn't understand your proposal correctly.

About the JSON-LD issue, I think it is a matter of taste. I think that this option has a better potential for adoption by web developpers over other RDF serialization languages (RDF/XML, TURTLE, RDFa, ...) because JSON is intensively used in web applications.

**Comment by fgalan** [comment]
More than 2 years have passed since this issue was updated last time. I think it can be closed.
